elastalert/elastalert.py

```diff
@@ -868,7 +876,7 @@ def run_rule(self, rule, endtime, starttime=None):
         :return: The number of matches that the rule produced.
         """
         run_start = time.time()
-        self.thread_data.current_es = self.es_clients.setdefault(rule['name'], elasticsearch_client(rule))
```

Comment:
`setdefault` results in the correct behaviour, but because it takes a concrete default value, `elasticsearch_client` is getting called every run and the result is disregarded.  From what I see `dictionary` does not have a lazy option.

Reply:
Yes, I see where this will help improve efficiency. I think a unit test that proves the same elasticsearch_client instance is returned from the new getter function, when called twice with the same rule, would be good unit coverage.

Reply:
This is especially bad for us running in AWS because inside `Auth` we call `get_credentials` which is an HTTP call to the Ec2 metdata service:

https://github.com/jertel/elastalert2/blob/d455e55e9180a709b5eabef1473d81b97a525b36/elastalert/auth.py#L59